Separate unknown -webkit-prefixed pseudo-elements from known

annevk opened the pull request:
#### 405fb5beffd040a63e4da9874d280e4e783c2aa7
<pre>
Separate unknown -webkit-prefixed pseudo-elements from known
<a href="https://bugs.webkit.org/show_bug.cgi?id=266947">https://bugs.webkit.org/show_bug.cgi?id=266947</a>

Reviewed by Tim Nguyen.

All UserAgentPartIds are moved into CSSPseudoSelectors.json and then
generated from there from now on.

This in combination with adding the WebKitUnknown pseudo-element
identifier corrects @supports so it no longer claims certain
-webkit-prefixed pseudo-elements are unsupported when they are in fact
supported.

Existing PseudoElement::UserAgentPart conditionals are amended to
account for PseudoElement::WebKitUnknown.

FIXMEs are added for areas that need further work.

* LayoutTests/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::pseudoId):
(WebCore::CSSSelector::parsePseudoElementName):
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::matchRecursively const):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isPseudoClassValidAfterPseudoElement):
(WebCore::CSSSelectorParser::containsUnknownWebKitPseudoElements):
* Source/WebCore/css/parser/MutableCSSSelector.h:
(WebCore::MutableCSSSelector::needsImplicitShadowCombinatorForMatching const):
* Source/WebCore/css/process-css-pseudo-selectors.py:
(InputValidator.validate_fields):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::constructFragmentsInternal):

No need for an else branch here due to the early return usage.

* Source/WebCore/html/shadow/UserAgentPartIds.cpp: Removed.
* Source/WebCore/html/shadow/UserAgentPartIds.h: Removed.

Canonical link: <a href="https://commits.webkit.org/272726@main">https://commits.webkit.org/272726@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/498d79c33545d14017cbdf7cfb11b8dab8b06966

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32832 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11582 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34697 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35445 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29656 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33760 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13925 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8765 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29104 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33277 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9766 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29320 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8504 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8647 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29276 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36776 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29824 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29680 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34766 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8766 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6734 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32640 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10456 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9376 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4234 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9388 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -394,7 +395,7 @@ inline bool CSSSelector::matchesPseudoElement() const
 inline bool CSSSelector::isWebKitCustomPseudoElement() const
 {
-    return pseudoElement() == PseudoElement::WebKitCustom || pseudoElement() == PseudoElement::WebKitCustomLegacyPrefixed;
+    return pseudoElement() == PseudoElement::WebKitCustom || pseudoElement() == PseudoElement::WebKitCustomLegacyPrefixed || pseudoElement() == PseudoElement::WebKitUnknown;
```
I wouldn't add `pseudoElement() == PseudoElement::WebKitUnknown` in here, especially with the new name. I'd just handle `PseudoElement::WebKitUnknown` as its own case in the only callsite of the function.

nt1m commented on Source/WebCore/css/CSSPseudoSelectors.json:
```diff
@@ -232,6 +232,162 @@
+        "-webkit-date-and-time-value": {
+            "status": "non-standard",
+            "user-agent-part": true,
+            "condition": "ENABLE(DATE_AND_TIME_INPUT_TYPES)"
```
nit: Here and below, can you move up the `condition` field to preserve alphabetic sorting?

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -324,14 +325,14 @@ std::optional<CSSSelector::PseudoElement> CSSSelector::parsePseudoElement(String
     if (!type) {
         // FIXME: Put all known UA parts in CSSPseudoSelectors.json and split out the unknown case (webkit.org/b/266947).
```
Remove this FIXME ?

nt1m commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -281,7 +281,7 @@ SelectorChecker::MatchResult SelectorChecker::matchRecursively(CheckingContext&
 
     if (context.selector->match() == CSSSelector::Match::PseudoElement) {
-        if (context.selector->isUserAgentPartPseudoElement()) {
+        if (context.selector->isUserAgentPartPseudoElement() || context.selector->pseudoElement() == CSSSelector::PseudoElement::WebKitUnknown) {
```
`PseudoElement::WebKitUnknown` should never match anything, we should fast track all the code below

annevk replied:
I would prefer making that as a separate change given the risk pointed out below. Unless you're sure there cannot be any web observable impact.

nt1m replied:
I'm pretty sure in this case that there's no web impact. With the current patch, selector matching will fail anyway, but it will go through the work of navigating shadow roots and stuff when it doesn't really need to. We can just fast track this and fail the matching early.

nt1m commented on Source/WebCore/css/parser/CSSSelectorParser.cpp:
```diff
@@ -452,6 +454,8 @@ static bool isPseudoClassValidAfterPseudoElement(CSSSelector::PseudoClass pseudo
     case CSSSelector::PseudoElement::UserAgentPart:
     case CSSSelector::PseudoElement::UserAgentPartLegacyAlias:
+    // FIXME: Should WebKitUnknown be included here?
+    case CSSSelector::PseudoElement::WebKitUnknown:
```
We should make this use the `default:` case, chaining an unknown webkit pseudo element and a pseudo class should not be allowed.

annevk replied:
I tend to agree, but that would be a change in behavior: http://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cstyle%3E%3A%3A-webkit-asdf%3Afocus%2C%20body%20%7B%20background%3Ablue%20%7D%20%3C%2Fstyle%3E

I'd rather make that change separately from the @supports case, with its own tests.

nt1m replied:
Seems fine to remove as a followup.

nt1m commented on Source/WebCore/css/parser/MutableCSSSelector.h:
```diff
@@ -112,12 +112,13 @@ inline bool MutableCSSSelector::needsImplicitShadowCombinatorForMatching() const
     return match() == CSSSelector::Match::PseudoElement
         && (pseudoElement() == CSSSelector::PseudoElement::UserAgentPart
+            || pseudoElement() == CSSSelector::PseudoElement::UserAgentPartLegacyAlias
+            || pseudoElement() == CSSSelector::PseudoElement::WebKitUnknown
```
I would just remove this case. We don't need to look into the shadow DOM for webkit unknown pseudos.

nt1m replied:
I still think we should remove this case, this is used to prepend implicit selectors to unknown ::-webkit pseudos, ... which shouldn't matter anyway if they don't match.

nt1m commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -734,6 +742,83 @@ def write_pseudo_argument_check_function(self, writer, function_name, enum_name,
 
+class UserAgentPartsGenerator():
+    def __init__(self, data):
+        self.data = data
```
You're not using this, I would just remove it.

nt1m commented on Source/WebCore/style/RuleSet.cpp:
```diff
@@ -189,6 +189,7 @@ void RuleSet::addRule(RuleData&& ruleData, CascadeLayerIdentifier cascadeLayerId
             switch (selector->pseudoElement()) {
             case CSSSelector::PseudoElement::UserAgentPart:
             case CSSSelector::PseudoElement::UserAgentPartLegacyAlias:
+            case CSSSelector::PseudoElement::WebKitUnknown:
```
I would just remove this case and skip to the default break;

nt1m reviewed (commented):
You'll need to register your new C++ file like https://searchfox.org/wubkat/rev/e0c29e7e482a4a452c92c394677358659c71eace/Source/WebCore/CMakeLists.txt#2555 to fix the GTK build fwiw

nt1m commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -734,6 +742,83 @@ def write_pseudo_argument_check_function(self, writer, function_name, enum_name,
+    def filter_data(self, data):
+        parts = {}
+        for part_name, part_data in data.items():
+            if "user-agent-part" in part_data or "user-agent-part-string" in part_data:
```
nit: single quotes

nt1m commented on Source/WebCore/css/parser/CSSSelectorParser.cpp:
```diff
@@ -287,7 +287,9 @@ static OptionSet<CompoundSelectorFlag> extractCompoundFlags(const MutableCSSSele
-    if (parserMode == UASheetMode && simpleSelector.pseudoElement() == CSSSelector::PseudoElement::UserAgentPart)
+    if (parserMode == UASheetMode
+        && (simpleSelector.pseudoElement() == CSSSelector::PseudoElement::UserAgentPart
+        || simpleSelector.pseudoElement() == CSSSelector::PseudoElement::WebKitUnknown))
```
Given we control the UA stylesheet, I don't expect us to have unknown -webkit- pseudo elements in there. We should also remove this special case.

nt1m reviewed (approved):
r=me minus the two usages that can be removed.

nt1m commented on Source/WebCore/css/CSSPseudoSelectors.json:
```diff
@@ -20,6 +20,10 @@
             "user-agent-part": [
                 "For pseudo-elements only. Whether the pseudo-element represents an element in an user agent shadow tree.",
                 "They internally map to `PseudoElement::UserAgentPart` or `PseudoElement::UserAgentPartLegacyAlias` for aliases."
```
Can we add a line saying this will generate code for `UserAgentPartIds`?

mcatanzaro commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -122,6 +122,7 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_INPUT_TYPE_DATETIMELOCAL PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_INPUT_TYPE_MONTH PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_INPUT_TYPE_TIME PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_INPUT_TYPE_WEEK PRIVATE ON)
+WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_DATE_AND_TIME_INPUT_TYPES ON)
```
You're passing `ON` where it should be `PUBLIC` or `PRIVATE`. Please alphabetize it and also make it `PRIVATE ON`.

mcatanzaro commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -2226,6 +2226,10 @@ if (USE_LIBWEBRTC)
+endif ()
 if (HAVE_OS_DARK_MODE_SUPPORT)
     set(FEATURE_DEFINES_WITH_SPACE_SEPARATOR "${FEATURE_DEFINES_WITH_SPACE_SEPARATOR} HAVE_OS_DARK_MODE_SUPPORT")
 endif ()
```
I think the only problem here is `ENABLE_DATE_AND_TIME_INPUT_TYPES` is not a real WebKit feature because it is not declared in WebKitFeatures.cmake (or any port-specific Options*.cmake). So yeah, just add `ENABLE_DATE_AND_TIME_INPUT_TYPES` to WebKitFeatures.cmake and this problem should go away. I think. Right?

Then I would add a different FIXME above this `if (HAVE_OS_DARK_MODE_SUPPORT)` block complaining that it is messy. WebKit features don't start with `HAVE_` so it's harder to decide what to do about that one.

annevk replied:
I do worry a bit that properly defining it would result in a bunch of new failures, but maybe not. I guess I'll experiment a bit more tomorrow. Maybe also create a separate PR just for that change.

This should have been essentially a no-op change so it's rather weird to now be stuck.

mcatanzaro commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -2546,14 +2550,16 @@ ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/CSSPrimitiveValueMapp
     VERBATIM)
 list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/SelectorPseudoClassAndCompatibilityElementMap.cpp)
 list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/SelectorPseudoElementMap.cpp)
+list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/UserAgentPartIds.cpp)
+list(APPEND WebCore_SOURCES ${WebCore_DERIVED_SOURCES_DIR}/UserAgentPartIds.h)
 ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/CSSSelector.cpp CSSSelectorEnums.h)
 ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/CSSSelector.cpp CSSSelectorInlines.h)
 ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES(${WEBCORE_DIR}/css/parser/CSSSelectorParser.cpp CSSSelectorInlines.h)
```
We need to guarantee that UserAgentPartIds.h is generated before attempting to compile any source files that include it, or they'll fail to build if you get unlucky. For now, use ADD_SOURCE_WEBCORE_DERIVED_DEPENDENCIES() for this. This macro is broken and does nothing, but at least it will document the dependency. I'll continue to investigate what we ought to do in https://bugs.webkit.org/show_bug.cgi?id=266984

annevk replied:
There's quite a few files that include it. I'll make an attempt, but it won't be pretty and I'm not entirely convinced it will scale long term.

mcatanzaro replied:
ack, it's awful and we need something better, but we can deal with it via https://bugs.webkit.org/show_bug.cgi?id=266984

nt1m commented on Source/WebCore/css/CSSPseudoSelectors.json:
```diff
@@ -19,7 +19,12 @@
                 "For pseudo-elements only. Whether the pseudo-element represents an element in an user agent shadow tree.",
-                "They internally map to `PseudoElement::UserAgentPart` or `PseudoElement::UserAgentPartLegacyAlias` for aliases."
+                "They internally map to `PseudoElement::UserAgentPart` or `PseudoElement::UserAgentPartLegacyAlias` for aliases.",
+                "Also causes an entry in UserAgentPartIds to be created."
```
```suggestion
                "Also creates an entry in UserAgentPartIds."
```
nit: "causes" reads as if it's a negative side-effect.

nt1m commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -792,6 +805,81 @@ def write_pseudo_argument_check_function(self, writer, function_name, enum_name,
+                writer.write('#if {}'.format(part_data['condition']))
+            writer.write('const AtomString& {}();'.format(format_name_for_function(part_name)))
+            if 'condition' in part_data:
+                writer.write('#endif')
```
Please use f-strings once you rebase on top of https://github.com/WebKit/WebKit/commit/3cd087d87fd5299b8c7f544ff2d377decfbc408e (which also has renamed `condition` to `conditional`)

nt1m commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -792,6 +805,81 @@ def write_pseudo_argument_check_function(self, writer, function_name, enum_name,
+            writer.write('{')
+            with writer.indent():
+                writer.write('static MainThreadNeverDestroyed<const AtomString> {}("{}"_s);'.format(part_function_name, part_name))
+                writer.write('return {};'.format(part_function_name))
```
ditto here

nt1m commented on Source/WebCore/css/parser/MutableCSSSelector.h:
```diff
@@ -122,7 +124,8 @@ inline bool MutableCSSSelector::needsImplicitShadowCombinatorForMatching() const
             || pseudoElement() == CSSSelector::PseudoElement::Part
             || pseudoElement() == CSSSelector::PseudoElement::Slotted
-            || pseudoElement() == CSSSelector::PseudoElement::UserAgentPartLegacyAlias);
+            || pseudoElement() == CSSSelector::PseudoElement::UserAgentPartLegacyAlias
+            || pseudoElement() == CSSSelector::PseudoElement::WebKitUnknown);
```
Can we move down `pseudoElement() == CSSSelector::PseudoElement::UserAgentPart` to have some alphabetic ordering?

annevk replied:
Does that all compile down to the same thing? I thought these might be ordered this way for perf reasons.

nt1m replied:
We should probably add a comment if that's the case. The temptation of re-ordering is easy. 

But I doubt it makes a noticeable difference tbh

annevk replied:
I'm going to leave this as-is for now. There's already enough subtle changes in this patch. And there's a FIXME for this function anyway so once that's addressed this can be addressed at the same time. (Or this change simply gets undone.)

nt1m replied:
Sounds good!

annevk committed:
Separate unknown -webkit-prefixed pseudo-elements from known
https://bugs.webkit.org/show_bug.cgi?id=266947

Reviewed by Tim Nguyen.

All UserAgentPartIds are moved into CSSPseudoSelectors.json and then
generated from there from now on.

This in combination with adding the WebKitUnknown pseudo-element
identifier corrects @supports so it no longer claims certain
-webkit-prefixed pseudo-elements are unsupported when they are in fact
supported.

Existing PseudoElement::UserAgentPart conditionals are amended to
account for PseudoElement::WebKitUnknown.

FIXMEs are added for areas that need further work.

* LayoutTests/TestExpectations:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::pseudoId):
(WebCore::CSSSelector::parsePseudoElementName):
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::matchRecursively const):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isPseudoClassValidAfterPseudoElement):
(WebCore::CSSSelectorParser::containsUnknownWebKitPseudoElements):
* Source/WebCore/css/parser/MutableCSSSelector.h:
(WebCore::MutableCSSSelector::needsImplicitShadowCombinatorForMatching const):
* Source/WebCore/css/process-css-pseudo-selectors.py:
(InputValidator.validate_fields):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::constructFragmentsInternal):

No need for an else branch here due to the early return usage.

* Source/WebCore/html/shadow/UserAgentPartIds.cpp: Removed.
* Source/WebCore/html/shadow/UserAgentPartIds.h: Removed.

Canonical link: https://commits.webkit.org/272726@main
